src/Config.hs

```diff
@@ -40,50 +47,286 @@ newtype MilliSeconds = MilliSeconds { unMilliSeconds :: Int }
-    , "Inherit env:    " ++ (show $ oInheritEnv opts)
-    , "Base delay:     " ++ (show . unMilliSeconds $ oRetryBaseDelay opts)
-    , "Retry attempts: " ++ (show $ oRetryAttempts opts)
-    , "Log-level:      " ++ (show $ oLogLevel opts)
-    , "Use PATH:       " ++ (show $ oUsePath opts)
-    ]
-
--- | Behavior flags that we allow users to set via environment variables.
--- This type is internal to the workings of this module. It is used as an
--- intermediate value to get optparse-applicative to play nice with environment
--- variables as used for behavior flags. All flags are off by default.
-data EnvFlags = EnvFlags
-  { efConnectTls :: Bool
-  , efValidateCerts :: Bool
-  , efInheritEnv :: Bool
-  , efUsePath :: Bool
+    [ "Host:           " ++ showSpecifiedString (oVaultHost opts)
+    , "Port:           " ++ showSpecified (oVaultPort opts)
+    , "Addr            " ++ showSpecifiedString (oVaultAddr opts)
+    , "Token:          " ++ maybe "Unspecified" (const "*****") (oVaultToken opts)
+    , "Secret file:    " ++ showSpecifiedString (oSecretFile opts)
+    , "Command:        " ++ showSpecifiedString (oCmd opts)
+    , "Arguments:      " ++ showSpecified (oArgs opts)
+    , "Use TLS:        " ++ showSpecified (oConnectTls opts)
+    , "Validate certs: " ++ showSpecified (oValidateCerts opts)
+    , "Inherit env:    " ++ showSpecified (oInheritEnv opts)
+    , "Base delay:     " ++ showSpecified (unMilliSeconds <$> oRetryBaseDelay opts)
+    , "Retry attempts: " ++ showSpecified (oRetryAttempts opts)
+    , "Log-level:      " ++ showSpecified (oLogLevel opts)
+    , "Use PATH:       " ++ showSpecified (oUsePath opts)
+    ] where
+      showSpecified :: Show a => Maybe a -> String
+      showSpecified (Just x) = show x
+      showSpecified Nothing = "Unspecified"
+      showSpecifiedString = fromMaybe "Unspecified"
+
+-- | Gets the option from a Maybe, ```name``` should only be used for debugging
+-- purposes, as ```isOptionsComplete``` should verify that every property that
+-- does not have a default value, is present in the options.
+getOptionsValue :: (Options Validated Completed -> Maybe c) -> Options Validated Completed -> c
+getOptionsValue f opts=
+    fromMaybe (error "Complete options has an unknown key") (f opts)
+
+-- | The possible errors that can occur in the construction of an Options datatype
+data OptionsError
+  = UnspecifiedValue String -- ^ A value is missing and no default is specified
+  | InvalidAddrFormat String -- ^ The format of the address is invalid
+  | UnknownScheme String -- ^ The scheme of the address is invalid, e.g. ftp://
+  | NonNumericPort String -- ^ The port of the address is not an valid integer
+  | HostPortSchemeAddrMismatch Bool String Int String
+      -- ^ The useTls, host, port and scheme do not match the provided address
+
+
+
+
+instance Show OptionsError where
+  show (UnspecifiedValue s) = "The option " ++ s ++ " is required but not specified"
```

Comment:
Sort of related: when we detect mulitple errors (awesome that we detect mulitple errors :clap:), we render them like:

```
[ERROR] [The option Command is required but not specified,The option Secret file is required but not specified]
```

Can we split that:

```
[ERROR] The option Command is required but not specified
[ERROR] The option Secret file is required but not specified
```

Reply:
This has been updated.